website/blog/2023-09-06-optimize-try-cast.mdx

```diff
@@ -0,0 +1,114 @@
+
+1. Folly library throw.
+2. cast function catch the exception and convert it to Velox exception and throw it in Conversions.h.
+3. CastToPrimitive catch the exception and throw it again with a different specific detailed message.
+4. Another throw happens inside applyToSelectedNoThrow, which used to catch that error and call setError which calls
+ toVeloxException which rethrow to check the exception type.
+
+Those are addressed and avoided using the following:
+
+1. When the input is an empty string, avoid calling folly by directly pre-checking if the input is empty.
+2. Remove the catch and re-throw from Conversions.h
+3. Introduce setVeloxExceptionError, which can be used to set the error directly in evalContext with out any throw (does not call toVeloxException).
+4. Optimize applyToSelectedNoThrow to call setVeloxExceptionError if it catches Velox exception.
+
+With all those changes throwing exceptions are completely avoided when casting empty strings. This takes the runtime down to 382.07ms ,
+but its still much higher than 2.15ms
+
+
+
+** Velox exception construction overhead.**
+
+The construction of Velox exceptions is expensive, even when there is no throw at all! Luckily this can be avoided for try_cast, since
+the output can be directly set to null without having to use the errorVector to track errors. By doing so the benchmark runtime goes
+down to  1.24ms.
+
+--------------------------------------------------------------------------------------------
+
+
+### Follow up tasks
+After all the changes we have the following performance numbers for other pattern of similar expressions
+(much better than before but still can be optimized a lot).
+```
+try_cast(empty_string_col as int)                     1.24ms    808.79
+
+try_cast(invalid_string_col as int)                  393.61ms     2.54
+
+try(cast(empty_string_col as int))                   375.82ms     2.66
+
+try(cast(invalid_string_col as int))                767.74ms      1.30
+```
+
+All of those should be possible to optimize to have the same runtime cost of the first expression 1.24ms.
+
+To do that two thing are needed:
+
+
+**1) Tracking errors for try, should not require constructing exceptions**
+
+The way errors are tracked when evaluating a try expression is by setting values in an  ErrorVector; which is a vector of VeloxException pointers.
+This forces the construction of a Velox exception for each row,  but that is not needed (for try expressions) since only row numbers need to be
+tracked to be converted eventually to nulls, but not the actual errors.
+
+This can be changed such that errors are tracked using a selectivity vector. Its worth noting that for other expressions such as conjunct
+expressions this tracking is needed, hence we need to distinguish between both.
+
+This would not just optimize try(cast(empty_string_col as int)) but it is also important to optimize any expression try(x) where x throws for
```

Comment:
perhaps, This would help optimize any try(x) expression where x throws for large number of rows.